File: core/amber/src/main/python/core/storage/document_factory.py
# Licensed to the Apache Software Foundation (ASF) under one
# or more contributor license agreements.  See the NOTICE file
# distributed with this work for additional information
# regarding copyright ownership.  The ASF licenses this file
# to you under the Apache License, Version 2.0 (the
# "License"); you may not use this file except in compliance
# with the License.  You may obtain a copy of the License at
#
#   http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing,
# software distributed under the License is distributed on an
# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
# KIND, either express or implied.  See the License for the
# specific language governing permissions and limitations
# under the License.

import typing
import urllib
from urllib.parse import urlparse

from typing import Optional

from core.models import Schema, Tuple
from core.storage.iceberg.iceberg_catalog_instance import IcebergCatalogInstance
from core.storage.iceberg.iceberg_document import IcebergDocument
from core.storage.iceberg.iceberg_utils import (
    create_table,
    amber_tuples_to_arrow_table,
    arrow_table_to_amber_tuples,
    load_table_metadata,
)
from core.storage.model.virtual_document import VirtualDocument
from core.storage.storage_config import StorageConfig
from core.storage.vfs_uri_factory import VFSURIFactory, VFSResourceType


class DocumentFactory:
    """
    Factory class to create and open documents.
    Currently only iceberg documents are supported.
    """

    ICEBERG = "iceberg"

    @staticmethod
    def sanitize_uri_path(uri):
        """
        Matches the same implementation in our Scala codebase.
        urllib.parse.urlparse does not automatically unquote the URI, while
        java.net.URI.getPath does. Hence we need to explicitly
        unquote to decode percent-encoded characters, then sanitize.
        :param uri: Result of urllib.parse.urlparse(). Could be quoted.
        :return: Unquoted and sanitized format of uri.
        """
        return urllib.parse.unquote(uri.path).lstrip("/").replace("/", "_")

    @staticmethod
    def create_document(uri: str, schema: Schema) -> VirtualDocument:
        parsed_uri = urlparse(uri)
        if parsed_uri.scheme == VFSURIFactory.VFS_FILE_URI_SCHEME:
            _, _, _, resource_type = VFSURIFactory.decode_uri(uri)

            if resource_type in {VFSResourceType.RESULT}:
                storage_key = DocumentFactory.sanitize_uri_path(parsed_uri)

                iceberg_schema = Schema.as_arrow_schema(schema)

                create_table(
                    IcebergCatalogInstance.get_instance(),
                    StorageConfig.ICEBERG_TABLE_RESULT_NAMESPACE,
                    storage_key,
                    iceberg_schema,
                    override_if_exists=True,
                )

                return IcebergDocument[Tuple](
                    StorageConfig.ICEBERG_TABLE_RESULT_NAMESPACE,
                    storage_key,
                    iceberg_schema,
                    amber_tuples_to_arrow_table,
                    arrow_table_to_amber_tuples,
                )
            else:
                raise ValueError(f"Resource type {resource_type} is not supported")
        else:
            raise NotImplementedError(
                f"Unsupported URI scheme: {parsed_uri.scheme} for creating the document"
            )

    @staticmethod
    def open_document(uri: str) -> typing.Tuple[VirtualDocument, Optional[Schema]]:
        parsed_uri = urlparse(uri)
        if parsed_uri.scheme == "vfs":
            _, _, _, resource_type = VFSURIFactory.decode_uri(uri)

            if resource_type in {VFSResourceType.RESULT}:
                storage_key = DocumentFactory.sanitize_uri_path(parsed_uri)

                table = load_table_metadata(
                    IcebergCatalogInstance.get_instance(),
                    StorageConfig.ICEBERG_TABLE_RESULT_NAMESPACE,
                    storage_key,
                )

                if table is None:
                    raise ValueError("No storage is found for the given URI")

                amber_schema = Schema(table.schema().as_arrow())

                document = IcebergDocument(
                    StorageConfig.ICEBERG_TABLE_RESULT_NAMESPACE,
                    storage_key,
                    table.schema(),
                    amber_tuples_to_arrow_table,
                    arrow_table_to_amber_tuples,
                )
                return document, amber_schema
            else:
                raise ValueError(f"Resource type {resource_type} is not supported")
        else:
            raise NotImplementedError(
                f"Unsupported URI scheme: {parsed_uri.scheme} for opening the document"
            )

File: common/workflow-core/src/main/scala/org/apache/texera/amber/core/storage/DocumentFactory.scala
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

package org.apache.texera.amber.core.storage

import org.apache.texera.common.config.StorageConfig
import org.apache.texera.amber.core.storage.FileResolver.DATASET_FILE_URI_SCHEME
import org.apache.texera.amber.core.storage.VFSResourceType._
import org.apache.texera.amber.core.storage.VFSURIFactory.{VFS_FILE_URI_SCHEME, decodeURI}
import org.apache.texera.amber.core.storage.model._
import org.apache.texera.amber.core.storage.result.iceberg.IcebergDocument
import org.apache.texera.amber.core.tuple.{Schema, Tuple}
import org.apache.texera.amber.util.IcebergUtil
import org.apache.iceberg.catalog.TableIdentifier
import org.apache.iceberg.data.Record
import org.apache.iceberg.{Schema => IcebergSchema}

import java.net.URI

object DocumentFactory {

  val ICEBERG = "iceberg"

  private def sanitizeURIPath(uri: URI): String =
    uri.getPath.stripPrefix("/").replace("/", "_")

  private def resolveNamespace(resourceType: VFSResourceType.Value): String =
    resourceType match {
      case RESULT             => StorageConfig.icebergTableResultNamespace
      case CONSOLE_MESSAGES   => StorageConfig.icebergTableConsoleMessagesNamespace
      case RUNTIME_STATISTICS => StorageConfig.icebergTableRuntimeStatisticsNamespace
      case STATE              => StorageConfig.icebergTableStateNamespace
      case _ =>
        throw new IllegalArgumentException(s"Resource type $resourceType is not supported")
    }

  /**
    * Open a document specified by the uri for read purposes only.
    * @param fileUri the uri of the document
    * @return ReadonlyVirtualDocument
    */
  def openReadonlyDocument(fileUri: URI): ReadonlyVirtualDocument[_] = {
    fileUri.getScheme match {
      case DATASET_FILE_URI_SCHEME => new DatasetFileDocument(fileUri)
      case "file"                  => new ReadonlyLocalFileDocument(fileUri)
      case unsupportedScheme =>
        throw new UnsupportedOperationException(
          s"Unsupported URI scheme: $unsupportedScheme for creating the ReadonlyDocument"
        )
    }
  }

  /**
    * Create a document for storage specified by the uri.
    * This document is suitable for storing structural data, i.e. the schema is required to create such document.
    * @param uri the location of the document
    * @param schema the schema of the data stored in the document
    * @return the created document
    */
  def createDocument(uri: URI, schema: Schema): VirtualDocument[_] = {
    uri.getScheme match {
      case VFS_FILE_URI_SCHEME =>
        val (_, _, _, resourceType) = decodeURI(uri)
        val storageKey = sanitizeURIPath(uri)
        val namespace = resolveNamespace(resourceType)

        val icebergSchema = IcebergUtil.toIcebergSchema(schema)
        IcebergUtil.createTable(
          IcebergCatalogInstance.getInstance(),
          namespace,
          storageKey,
          icebergSchema,
          overrideIfExists = true
        )
        val serde: (IcebergSchema, Tuple) => Record = IcebergUtil.toGenericRecord
        val deserde: (IcebergSchema, Record) => Tuple = (schema, record) =>
          IcebergUtil.fromRecord(record, IcebergUtil.fromIcebergSchema(schema))

        new IcebergDocument[Tuple](
          namespace,
          storageKey,
          icebergSchema,
          serde,
          deserde
        )
      case unsupportedScheme =>
        throw new UnsupportedOperationException(
          s"Unsupported URI scheme: $unsupportedScheme for creating the document"
        )
    }
  }

  /**
    * Check whether a document exists at the given URI without opening it.
    *
    * Returns true iff the underlying storage already has an entry for this
    * URI (e.g., an iceberg table at the resolved namespace + storage key).
    *
    * @throws UnsupportedOperationException if the URI scheme is not `vfs`.
    * @throws IllegalArgumentException if the resolved resource type has no
    *                                  iceberg namespace mapping.
    */
  def documentExists(uri: URI): Boolean = {
    uri.getScheme match {
      case VFS_FILE_URI_SCHEME =>
        val (_, _, _, resourceType) = decodeURI(uri)
        val storageKey = sanitizeURIPath(uri)
        val namespace = resolveNamespace(resourceType)
        IcebergCatalogInstance
          .getInstance()
          .tableExists(TableIdentifier.of(namespace, storageKey))

      case unsupportedScheme =>
        throw new UnsupportedOperationException(
          s"Unsupported URI scheme: $unsupportedScheme for checking document existence"
        )
    }
  }

  /**
    * Return the document at `uri`: when `reuseExisting` is set and a document
    * already exists there, open and return the existing one -- so a caller whose
    * output accumulates across re-runs (e.g. a LoopEnd port whose region
    * re-executes once per loop iteration) keeps the already-populated document
    * instead of clobbering it, since `createDocument` overrides any existing
    * document. Otherwise create it.
    *
    * `exists` / `open` / `create` default to this object's own `documentExists`
    * / `openDocument` / `createDocument`; they are parameterized only so the
    * create-or-reuse decision can be unit-tested without an iceberg backend.
    */
  def createOrReuseDocument(
      uri: URI,
      schema: Schema,
      reuseExisting: Boolean,
      exists: URI => Boolean = documentExists,
      open: URI => VirtualDocument[_] = (u: URI) => openDocument(u)._1,
      create: (URI, Schema) => VirtualDocument[_] = createDocument
  ): VirtualDocument[_] =
    if (reuseExisting && exists(uri)) open(uri) else create(uri, schema)

  /**
    * Open a document specified by the uri.
    * If the document is storing structural data, the schema will also be returned
    * @param uri the uri of the document
    * @return the VirtualDocument, which is the handler of the data; the Schema, which is the schema of the data stored in the document
    */
  def openDocument(uri: URI): (VirtualDocument[_], Option[Schema]) = {
    uri.getScheme match {
      case DATASET_FILE_URI_SCHEME => (new DatasetFileDocument(uri), None)
      case VFS_FILE_URI_SCHEME =>
        val (_, _, _, resourceType) = decodeURI(uri)
        val storageKey = sanitizeURIPath(uri)
        val namespace = resolveNamespace(resourceType)

        val table = IcebergUtil
          .loadTableMetadata(
            IcebergCatalogInstance.getInstance(),
            namespace,
            storageKey
          )
          .getOrElse(
            throw new IllegalArgumentException("No storage is found for the given URI")
          )

        val amberSchema = IcebergUtil.fromIcebergSchema(table.schema())
        val serde: (IcebergSchema, Tuple) => Record = IcebergUtil.toGenericRecord
        val deserde: (IcebergSchema, Record) => Tuple = (schema, record) =>
          IcebergUtil.fromRecord(record, IcebergUtil.fromIcebergSchema(schema))

        (
          new IcebergDocument[Tuple](
            namespace,
            storageKey,
            table.schema(),
            serde,
            deserde
          ),
          Some(amberSchema)
        )
      case unsupportedScheme =>
        throw new UnsupportedOperationException(
          s"Unsupported URI scheme: $unsupportedScheme for opening the document"
        )
    }
  }
}
